packages/test-base/src/androidAndroidTest/kotlin/io/realm/kotlin/test/shared/CopyFromRealmTests.kt

```diff
@@ -0,0 +1,755 @@
+        val copiedList = unmanagedObj.objectListField
+        assertEquals(5, copiedList.size)
+        copiedList.forEachIndexed { i, el ->
+            assertFalse(el.isManaged())
+            assertEquals((i + 1).toString(), el.stringField)
+        }
+    }
+
+    @Test
+    fun embeddedObjectLists() {
+        // Create object with list of 5 objects
+        val sample = EmbeddedParent().apply {
+            children = (1..5).map { i ->
+                EmbeddedChild(i.toString())
+            }.toRealmList()
+        }
+
+        val insertedObj = realm.writeBlocking {
+            copyToRealm(sample)
+        }
+
+        val unmanagedObj: EmbeddedParent = insertedObj.copyFromRealm()
+
+        // Close Realm to ensure data is decoupled from Realm
+        realm.close()
+
+        assertNotSame(insertedObj, unmanagedObj)
+        assertNotNull(unmanagedObj.children)
+        val copiedList = unmanagedObj.children
+        assertEquals(5, copiedList.size)
+        copiedList.forEachIndexed { i, el ->
+            assertFalse(el.isManaged())
+            assertEquals((i + 1).toString(), el.id)
+        }
+    }
+
+    @Test
+    fun primitiveSets() {
+        val type = Sample::class
+        val schemaProperties = type.realmObjectCompanionOrThrow().io_realm_kotlin_schema().properties
+        val fields: Map<String, KProperty1<*, *>> = type.realmObjectCompanionOrThrow().io_realm_kotlin_fields
+
+        // Dynamically set data on the Sample object
+        val originalObject = Sample()
+        schemaProperties.forEach { prop: RealmProperty ->
+            if (prop.type is SetPropertyType) {
+                val accessor: KMutableProperty1<BaseRealmObject, Any?> = fields[prop.name] as KMutableProperty1<BaseRealmObject, Any?>
+                val set: Set<Any?> = createPrimitiveSetData(prop, accessor)
+                accessor.set(originalObject, set)
+            }
+        }
+
+        // Round-trip object through `copyToRealm` and `copyFromRealm`.
+        val unmanagedCopy = realm.writeBlocking {
+            copyToRealm(originalObject).copyFromRealm()
+        }
+
+        // Close Realm to ensure data is decoupled from Realm
+        realm.close()
+
+        // Validate that all primitive list fields were round-tripped correctly.
+        schemaProperties.forEach { prop: RealmProperty ->
+            if (prop.type is SetPropertyType) {
+                val accessor: KMutableProperty1<BaseRealmObject, Any?> = fields[prop.name] as KMutableProperty1<BaseRealmObject, Any?>
+                val set: Set<Any?> = createPrimitiveSetData(prop, accessor)
+
+                if (prop.type.storageType == RealmStorageType.BINARY) {
+                    val copy = accessor.get(unmanagedCopy) as Set<ByteArray?>
+                    assertEquals(set.size, copy.size)
+                    copy.forEach { copiedValue: ByteArray? ->
+                        // Order is not guaranteed in the set when round-tripped through Core.
```

Comment:
You can just assert that you can remove each element and assert that the set is empty afterwards

Reply:
Not sure it would make things easier? The problem with ByteArray also manifests on `remove` as `HashSet` would use the equals implementation which doesn't work correctly.

Reply:
Maybe I am just a bit confused about the logic and or comment. I think it would be more clear with (if I understand it correctly):
```
copy.forEach { copiedValue ->
    assertTrue {
        set.removeIf {
            copiedValue == null || (it as ByteArray).contentEquals(copiedValue)
        }
    }
}
assertTrue { set.isEmpty() }
```
Maybe it is just the huge comment that confuses me 🤪

Reply:
I think I would just have spun an implementation around `removeIf` and verify that we removed all elements that matched a predicate from the `copiedValue`, but I guess it maybe becomes more tricky to do as `removeIf` could remove multiple elements in one go. Maybe I was also just confused about the bloated comment 🤣